### PR TITLE
OAuth Integration - GitHub and Google

### PR DESCRIPTION
**Pull request by avivey**
This change adds support for login via OAuth (https://github.com/KerbalStuff/KerbalStuff/issues/127)

Supports GitHub and Google, and adding more providers would not be too hard.

### CODE REVIEW

**Comment by avivey** [comment]
the "3300 lines deleted" comes from the static content being regenerated as single-line files... Without the last commit, there's around 40 total lines deleted.

**Comment by SirCmpwn** [comment]
Nice work! I'll review and deploy this tomorrow.

File: KerbalStuff/blueprints/login_oauth.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        good = False
+    if check_email_for_registration(email):
+        good = False
+
+    if good:
+        password = binascii.b2a_hex(os.urandom(99))
+        user = User(username, email, password)
+        user.confirmation = binascii.b2a_hex(os.urandom(20)).decode("utf-8")
+        db.add(user)
+        db.commit()  # Need to commit now to get an ID.
```
Comment (SirCmpwn) [inline KerbalStuff/blueprints/login_oauth.py]: Actually you can use `db.flush()` to get an ID

**Comment by SirCmpwn** [comment]
You need to make an alembic migration for your new table.

**Comment by SirCmpwn** [comment]
I'm also not a big fan of the UI on the profile page. Something more like this, perhaps:



I have the changes for this if you like it. There's also some weird padding in the button on the login dropdown.

Overall, I'm a fan. Nicely done!

File: KerbalStuff/blueprints/login_oauth.py (hunk 1 above)
Comment (avivey) [inline KerbalStuff/blueprints/login_oauth.py]: Should I use `flush()`? What's the difference?

Comment (SirCmpwn) [inline KerbalStuff/blueprints/login_oauth.py]: When you use commit, you actually create rows in the database. When you flush, you just flush the transaction, and SQL gives you the ID, but doesn't commit the rows. If something goes wrong after a flush, there isn't going to be bad/incomplete data in the database.

**Comment by avivey** [comment]
I'm no good when it comes to UI, so I'd take whatever you have ready.

**Comment by SirCmpwn** [comment]
```diff
diff --git a/templates/profile.html b/templates/profile.html
index d08e5ec..aa6d0c4 100644
--- a/templates/profile.html
+++ b/templates/profile.html
@@ -51,21 +51,6 @@ window.username = "{{profile.username}}";
                                     <span class="glyphicon glyphicon-link"></span>
                                 </div>
                                 <div class="timeline-label">
-                                    <h2 style="margin-bottom: -20px;">
-                                        <span class="text-muted">
-                                            KSP Forums:
-                                        </span>
-                                        <input type="text" class="form-control" name="ksp-forum-user" value="{%if profile.forumUsername %}{{ profile.forumUsername }}{%endif%}">
-                                    </h2>
-                                </div>
-                            </div>
-                        </div>
-                        <div class="timeline-entry">
-                            <div class="timeline-entry-inner">
-                                <div class="timeline-icon">
-                                    <span class="glyphicon glyphicon-link"></span>
-                                </div>
-                                <div class="timeline-label">
                                     <h2 style="margin-bottom: -10px;">
                                         <span class="text-muted">
                                             Twitter:
@@ -126,7 +111,7 @@ window.username = "{{profile.username}}";
 <div class="container lead">
     <div class="row">
         <div class="col-md-8">
-            <h1 title="Login">Login<small> Login Configuration (optional):</small></h1>
+            <h1 title="Login">Connected Accounts</h1>
         </div>
 
     </div>
@@ -135,47 +120,44 @@ window.username = "{{profile.username}}";
     <div class="info-list">
         <div class="container">
             <div class="row">
-              <div class="timeline-centered">
-              {% for provider in oauth_providers %}
-                {% set provider_icon = oauth_providers[provider].icon %}
-                {% set provider_full_name = oauth_providers[provider].full_name %}
-                {% set auth = oauth_providers[provider].has_auth %}
+                <div class="col-md-6">
+                  <div class="timeline-centered">
+                  {% for provider in oauth_providers %}
+                    {% set provider_icon = oauth_providers[provider].icon %}
+                    {% set provider_full_name = oauth_providers[provider].full_name %}
+                    {% set auth = oauth_providers[provider].has_auth %}
 
-                  <div class="timeline-entry">
-                    <div class="timeline-entry-inner">
-                      <div class="timeline-icon">
-                        <span class="glyphicon glyphicon-link"></span>
-                      </div>
-                      <div class="timeline-label">
-                        <h2 style="margin-top: -10px; margin-bottom: 0;">
-                          <span class="text-muted">
-                              {{provider_full_name}}:
-                          </span>
-                          <br>
-                          {% if auth %}
-                            Your account is already configured for login with {{provider_full_name}}.
-                            <button data-toggle="modal" data-target="#confirm-disconnect-{{provider}}" class="btn btn-primary">
-                              <span class="fa-stack fa">
-                                <i class="fa fa-ban fa-stack-2x text-danger"></i>
-                                <i class="fa fa-{{provider_icon}} fa-stack-1x"></i>
-                              </span>
-                              Disconnect Your {{provider_full_name}} Account
-                            </button>
-                          {% else %}
-                            <form action="/connect-oauth" method="POST">
-                              Be able to login using your {{provider_full_name}} account:
-                              <input type="hidden" name="provider" value="{{provider}}">
-                              <button type="submit" class="btn btn-primary">
-                                <i class="fa fa-{{provider_icon}}"></i> Connect Your {{provider_full_name}} Account
-                              </button>
-                            </form>
-                          {% endif %}
-                        </h2>
+                      <div class="timeline-entry">
+                        <div class="timeline-entry-inner">
+                          <div class="timeline-icon">
+                            <i class="fa fa-{{provider_icon}}"></i>
+                          </div>
+                          <div class="timeline-label">
+                            <h2 style="margin-top: -10px; margin-bottom: 0;">
+                              {% if auth %}
+                                <button data-toggle="modal" data-target="#confirm-disconnect-{{provider}}" class="btn btn-default btn-sm">
+                                  Disconnect your {{provider_full_name}} account
+                                </button>
+                              {% else %}
+                                <form action="/connect-oauth" method="POST">
+                                  <input type="hidden" name="provider" value="{{provider}}">
+                                  <button type="submit" class="btn btn-default btn-sm">
+                                    Connect your {{provider_full_name}} account
+                                  </button>
+                                </form>
+                              {% endif %}
+                            </h2>
+                          </div>
+                        </div>
                       </div>
-                    </div>
+                  {% endfor %}
                   </div>
-              {% endfor %}
-              </div>
+                </div>
+                <div class="col-md-6">
+                    <p>
+                        Connecting your account on another service allows you to log in with that service.
+                    </p>
+                </div>
             </div>
         </div>
     </div>
```

**Comment by avivey** [comment]
image
The timeline icon looks tiny small on my system (Chrome/windows).
Is there a trick to make it large and centered? Using `fa-2x` makes it uglier: 
image

**Comment by avivey** [comment]
It looks like the alembic migration needs to be empty?
https://github.com/KerbalStuff/KerbalStuff/blob/master/alembic/versions/1ccdaebac05_add_database_tracked_game_versions.py

**Comment by SirCmpwn** [comment]
Sorry, missed one:

```diff
diff --git a/styles/timeline.scss b/styles/timeline.scss
index 3cfb063..97e7c2a 100644
--- a/styles/timeline.scss
+++ b/styles/timeline.scss
@@ -145,7 +145,7 @@
     box-shadow: 0 0 0 5px #333;
     float: left;
 
-    .glyphicon {
+    .glyphicon, .fa {
         font-size: 22px;
         line-height: 40px;
     }
```

This fixes the icons being small.

As for the migration, run `alembic revision --autogenerate -m "Summary of changes"`

**Comment by avivey** [comment]
The weird padding looks like it's because `.nav .dropdown input.btn` has `margin-top:10px`, but playing with it only made it worse.

**Comment by SirCmpwn** [comment]
Well, it's not a big deal. Take care of the other issues and I'll look into the padding weirdness when I get time.

**Comment by avivey** [comment]
Updated:
- use `flush`() where possible
- add migration
- took your UI changes

**Comment by SirCmpwn** [comment]
Thanks!

**Comment by avivey** [comment]
:-)

**Comment by ddevault** [comment]
It works :tada:

**Comment by avivey** [comment]
:cool:

**Pull request merged**